### PR TITLE
add ARNIQA metric

### LINKED ISSUE #2887: Add ARNIQA metric
## 🚀 Feature

Add a No-Reference Image Quality Assessment (NR-IQA) metric - **ARNIQA** [(Agnolucci et al., 2024)](https://arxiv.org/abs/2310.14918)

### Motivation

ARNIQA evaluates the technical quality of an image without the need for a high-quality reference and demonstrates a higher correlation to human judgments and stronger generalization capabilities than popular state-of-the-art NR-IQA metrics. ARNIQA only requires a single forward pass of a ResNet50 + linear layer for inference, and will not bring any new dependencies.

### Pitch

ARNIQA module and functional support

### Additional context
I've already read the contributing guidelines of ```torchmetrics``` and added ARNIQA to a [popular IQA library](https://github.com/chaofengc/IQA-PyTorch/pull/158), so I would be happy to take care of the implementation.

### PR DESCRIPTION
**Pull request by LorenzoAgnolucci**
## What does this PR do?

Fixes #2887. Add [ARNIQA](https://arxiv.org/abs/2310.14918) functional and module support. 

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/torchmetrics/functional/image/arniqa.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+import torch
+import torch.nn.functional as F
+import torchvision
+from torch import Tensor, nn
+from typing_extensions import Literal
+
+from torchmetrics.utilities.imports import _TORCHVISION_AVAILABLE
+
+_available_regressor_datasets = {
```
Comment (Borda) [inline src/torchmetrics/functional/image/arniqa.py:31]: lets have these CAPITAL as they seems to be constants

File: src/torchmetrics/functional/image/arniqa.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+    def __init__(self, regressor_dataset: str = "koniq10k") -> None:
+        super().__init__()
+
+        if not _TORCHVISION_AVAILABLE:
+            raise ModuleNotFoundError(
+                "ARNIQA metric requires that torchvision is installed."
+                " Either install as `pip install torchmetrics[image]` or `pip install torchvision`."
+            )
+
+        valid_regressor_dataset = ("kadid10k", "koniq10k")
```
Comment (Borda) [inline src/torchmetrics/functional/image/arniqa.py:60]:
```suggestion
        valid_regressor_dataset = _available_regressor_datasets.keys()
```

File: src/torchmetrics/functional/image/arniqa.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+        self.regressor = nn.Linear(self.feat_dim * 2, 1)
+        self._load_weights()
+
+        self.encoder.eval()
+        self.regressor.eval()
+
+        for p in self.encoder.parameters():
+            p.requires_grad = False
+        for p in self.regressor.parameters():
+            p.requires_grad = False
```
Comment (Borda) [inline src/torchmetrics/functional/image/arniqa.py:83]:
```suggestion
        def _freeze(module):
            model.eval()
            for p in model.parameters():
                p.requires_grad = False

        _freeze(self.encoder)
        _freeze(self.regressor)
```
easier to follow

File: src/torchmetrics/functional/image/arniqa.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+        img_ds_f = self.encoder(img_ds)
+        img_ds_f = img_ds_f.view(-1, self.feat_dim)
+        img_ds_f = F.normalize(img_ds_f, dim=1)
+        f = torch.hstack((img_f, img_ds_f))
+
+        # Get the quality score
+        score = self.regressor(f)
+        score = self._scale_score(score)
+
+        return score
```
Comment (Borda) [inline src/torchmetrics/functional/image/arniqa.py:140]:
```suggestion
        return self._scale_score(score)
```

File: src/torchmetrics/functional/image/arniqa.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+        if regressor_dataset not in valid_regressor_dataset:
+            raise ValueError(
+                f"Argument `regressor_dataset` must be one of {valid_regressor_dataset}, but got {regressor_dataset}."
+            )
+
+        self.regressor_dataset = regressor_dataset
+        self.imagenet_norm_mean = [0.485, 0.456, 0.406]
+        self.imagenet_norm_std = [0.229, 0.224, 0.225]
+
+        encoder = torchvision.models.resnet50()
```
Comment (Borda) [inline src/torchmetrics/functional/image/arniqa.py:70]: can we have this as an argument so user for testing, we can use much smaller models?
ok, looking at https://github.com/miccunifi/ARNIQA/tree/weights/models so the available models are quite limited...

Comment (LorenzoAgnolucci) [reply src/torchmetrics/functional/image/arniqa.py:70]: The only pre-trained model available is ResNet50, that's why there is no argument for it

**Comment by LorenzoAgnolucci** [comment]
@Borda @lantiga @SkafteNicki @justusschock @stancld could you please let me know if I need to make any further changes to the PR?

**Review by SkafteNicki — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
@LorenzoAgnolucci I had to tweak bit the autocasting section since it was supported on older Torch versions, so wondering how important it is if you say it needed we just restrict that this metric will be supported only on newer Torch versions...

**Comment by LorenzoAgnolucci** [comment]
> @LorenzoAgnolucci I had to tweak bit the autocasting section since it was supported on older Torch versions, so wondering how important it is if you say it needed we just restrict that this metric will be supported only on newer Torch versions...

It's not that important, so your modifications are completely fine!
